src/Uno.UI.Runtime.Skia.X11/LinuxFilePickerExtension.cs

```diff
@@ -43,133 +44,161 @@ public async Task<IReadOnlyList<StorageFile>> PickMultipleFilesAsync(Cancellatio
+					(false, true) => ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_TITLE_DIRECTORY_SINGLE"),
+					(false, false) => ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_TITLE_FILE_SINGLE")
+				},
+				options: new Dictionary<string, VariantValue>
+				{
+					{ "handle_token", handleToken },
+					{ "accept_label", string.IsNullOrEmpty(picker.CommitButtonTextInternal) ? ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_ACCEPT_LABEL") : picker.CommitButtonTextInternal },
+					{ "multiple", multiple },
+					{ "directory", directory },
+					{ "filters", GetPortalFilters() },
+					{ "current_folder", new Array<byte>(Encoding.UTF8.GetBytes(PickerHelpers.GetInitialDirectory(picker.SuggestedStartLocationInternal)).Append((byte)'\0')) }
+				});
+
+			if (actualRequestPath != requestPath)
 			{
 				if (this.Log().IsEnabled(LogLevel.Error))
 				{
-					this.Log().Error($"File picker failed to receive a response.", e);
+					this.Log().Error($"{nameof(chooser.OpenFileAsync)} returned a path '{actualRequestPath}' that is different from supplied handle_token -based path '{requestPath}'");
 				}
 
-				tcs.TrySetResult(Array.Empty<string>());
-			});
+				return ImmutableList<string>.Empty;
+			}
 
-			var title = (multiple, directory) switch
-			{
-				(true, true) => ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_TITLE_DIRECTORY_MULTIPLE"),
-				(true, false) => ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_TITLE_FILE_MULTIPLE"),
-				(false, true) => ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_TITLE_DIRECTORY_SINGLE"),
-				(false, false) => ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_TITLE_FILE_SINGLE")
-			};
-			var window = "x11:" + host.RootX11Window.Window.ToString("X", CultureInfo.InvariantCulture);
-			var requestPath2 = await fileChooser.OpenFileAsync(window, title, new Dictionary<string, object>
-			{
-				{ "handle_token", handleToken },
-				{ "accept_label", string.IsNullOrEmpty(picker.CommitButtonTextInternal) ? ResourceAccessor.GetLocalizedStringResource("FILE_PICKER_ACCEPT_LABEL") : picker.CommitButtonTextInternal },
-				{ "multiple", multiple },
-				{ "directory", directory },
-				{ "filters", GetPortalFilters(picker.FileTypeFilterInternal) },
-				{ "current_folder", Encoding.UTF8.GetBytes(PickerHelpers.GetInitialDirectory(picker.SuggestedStartLocationInternal)).Append((byte)'\0') }
-			});
+			var (response, results) = await responseTcs.Task;
 
-			if (requestPath != requestPath2)
+			if (!Enum.IsDefined(typeof(Response), response))
 			{
 				if (this.Log().IsEnabled(LogLevel.Error))
 				{
-					this.Log().Error($"We are waiting at a wrong path {requestPath} instead of {requestPath2}");
+					this.Log().Error($"FileChooser returned an invalid response number {response}.");
 				}
 
-				tcs.TrySetResult(Array.Empty<string>());
+				return ImmutableList<string>.Empty;
 			}
 
-			return await tcs.Task;
+
+			if ((Response)response is not Response.Success)
```

Comment:
nit: Would not unboxing `response` earlier, prevent unboxing the same object in lines 180 and 182?